### ISSUE TITLE
Machine-readable documentation

### ISSUE THREAD

**Issue by astoff** [pgf-tikz/pgf#794]
It would be great if the PGF manual was easier to parse in order to extract the basic details about each command and key. This, in turn, would be useful for a text editor to provide completions, docstrings, and links to the full manual.

Currently, it is easy to extract from the manual a list of commands, environments, keys, libraries, etc. It's feasible (but not completely reliable) to get the argument signature of each command and key. And it's impossible to get a short docstring (the explanations in the manual are not such that the first sentence alone gives a quick overview).

This is a rather vague suggestion, and there are many approaches that could help. One idea would be to extend the manual macros so that, at each compilation, a list of all these objects and their interesting properties is produced (and dumped as json or something similar).

**Comment by hmenke** [comment]
Do you have a particular use-case in mind?  It seems like this is just going to be a huge amount of work for me with no real benefit to the average PGF/Ti*k*Z user.  If you need this for interfacing with other languages than TeX, I have to disappoint you.  The development audience of PGF/Ti*k*Z is only TeX users.

**Comment by hmenke** [comment]
Also, TeXstudio for example already has autocompletion for Ti*k*Z:
https://github.com/texstudio-org/texstudio/blob/master/completion/tikz.cwl

**Comment by hmenke** [comment]
There was a related question on TeX.SX once:
https://tex.stackexchange.com/questions/84753/summary-of-tikz-commands

**Comment by astoff** [comment]
> Do you have a particular use-case in mind?

My use-case is autocompletion inside a TeX editor, just like TeXstudio. But I would like to have complete, easy to update data, and I would like to have docstrings and links to the manual. In fact, I have done something in this direction for the autocompletion library that I'm writing; you can see the data file I generated [here](https://github.com/astoff/digestif/blob/master/digestif-data/tikz.lua).

Now, the scripts I used to generate those data files are really ugly, and need some manual intervention. So it would be great if PGF/Ti_k_Z itself could either provide this data, or make it easier for others to extract.

>  It seems like this is just going to be a huge amount of work for me with no real benefit to the average PGF/Ti_k_Z user

Granted this is a lot of work. And I'm not hoping this is going to just get done somehow. But perhaps you could come up with a systematic way to document command and key attributes, and use it for new code. Then perhaps the existing codebase will catch up eventually.

The benefit to the average PGF user is clear. The manual is over 1000 pages long, and although it's very nice to read, it can be hard to quickly find little details, which is where autocompletion helps.

**Comment by astoff** [comment]
Here's one pretty down-to-earth idea that would help. In the manual, the `{command}` environment takes one argument with the command signature. It's usually something easy to parse, like this:
``` latex
\begin{command}{\usepgfmodule\marg{module names}}
```
But sometimes there is random markup inside this argument, like `\texttt` or `|{|`. If it was possible to get rid of all this visual markup, and replace it by some "machine-readable" logical markup, it would help a lot.

**Comment by hmenke** [comment]
It's not possible to get rid of the additional markup because that is just how TeX works.  It should be obvious why
```latex
\begin{math-operators}{{}}{array}{}
```
will not display correctly and
```latex
\begin{math-operators}{\char`\{}{\char`\}}{array}{}
```
is necessary.  You could pack it in a box and walk the resulting node list in LuaTeX, but that would only work at runtime and not statically from source, which what you want I guess.

**Comment by astoff** [comment]
Well, one could easily replace ``\char`\{`` with a macro, say `\leftbracechar`; likewise, one could define macros so that
``` latex
\begin{command}{\pgfextra \meta{code} \texttt{\char`\\endpgfextra}}
```
can be rewritten as, say,
``` latex
\begin{command}{\pgfextra \meta{code} \cmd{endpgfextra}}
```
and so on. Note that there is already the `\marg` macro and variants, but some places in the manual do manual markup like ``{\char`\{}\meta{argument name}{\char`\}}``. This variety of syntaxes makes the manual hard to parse.

By cleaning things up in this way (only inside the `{command}` argument and similar places), it would be much more reliable to infer command signatures from the manual.

**Comment by hmenke** [comment]
This sounds to me like you are not parsing the manual at all, but instead trying to regex-match your way through.  This will not and cannot work because TeX is not a regular language.  Not sure whether I want to support poor matchers at the expense of reworking the markup of thousands of pages of manual.

**Comment by astoff** [comment]
> This sounds to me like you are [...] trying to regex-match your way through

Exactly! But there is a reason for this. If I wanted to reconstruct the output of the manual letter by letter, I would just run latex. Instead, I want to collect semantics that are only accessible in the logical markup.

This is exactly the same reason why we write `\section{Such and such}` and not `{\par\bg\Large 1. Such and such}`. The former has more information, although the PDF output is the same; for instance, I can search for `\\[a-z]*section\b` in my text editor to jump to the next (sub)section. What I want to do here with the manual is in the same spirit: to harness the extra information that the logical markup contains — and yes, by regexping it!

So, anyway, this is the proposal: replace all the manual formatting in the argument of `{command}` by purely logical markup (as I indicated in [my comment above](https://github.com/pgf-tikz/pgf/issues/794#issuecomment-560096330). Note that this is already the case in 95% of the instances.

Of course I am the one interested in this, so let me know if you would accept this change, if I made it.

**Comment by hmenke** [comment]
In general I am happy to accept changes.  Pretty much all the pull requests so far were accepted in one way or another.

**Comment by hmenke** [comment]
@astoff How are you coming along with this?  Have you made any progress?

**Comment by astoff** [comment]
The command list I generated a while ago (with some manual work) is still good enough. I might work on this issue when I need to update my data files.
